### PR TITLE
Find CUDA libraries with NVHPC package

### PR DESCRIPTION
**Pull request by tcojean**
As the title says. There might be other fixes piggy-backed in this PR.

### CODE REVIEW

File: cmake/cuda.cmake [hunk 1]
```diff
@@ -21,13 +21,44 @@ cas_variable_cuda_architectures(GINKGO_CUDA_ARCH_FLAGS
     ARCHITECTURES ${GINKGO_CUDA_ARCHITECTURES}
     UNSUPPORTED "20" "21")
 
+if (${CMAKE_CXX_COMPILER_ID} MATCHES "PGI|NVHPC")
+    find_package(NVHPC REQUIRED
```
Comment (tcojean) [inline cmake/cuda.cmake:25]: Does this need to be required in this case, or we silently disable CUDA if not found?

File: include/ginkgo/core/stop/combined.hpp [hunk 2]
```diff
@@ -125,7 +125,6 @@ std::shared_ptr<const CriterionFactory> combine(FactoryContainer&& factories)
     switch (factories.size()) {
     case 0:
         GKO_NOT_SUPPORTED(nullptr);
-        return nullptr;
```
Comment (yhmtsai) [inline include/ginkgo/core/stop/combined.hpp:128]: the first condition in default can also delete the `return nullptr`

File: omp/distributed/partition_kernels.cpp [hunk 3]
```diff
@@ -60,6 +60,8 @@ void build_starting_indices(std::shared_ptr<const DefaultExecutor> exec,
     auto size_per_thread =
         static_cast<size_type>(ceildiv(num_ranges, num_threads));
     vector<LocalIndexType> local_sizes(num_parts * num_threads, 0, {exec});
+    // TODO: PGI/NVHPC(22.7) doesn't like reduction with references
+    int tmp_reduction = 0;
```
Comment (yhmtsai) [inline omp/distributed/partition_kernels.cpp:64]: should we do it in this pr? We do not support omp with PGI/NVHPC, so we can just mention somewhere without changing it

**Review by yhmtsai — APPROVED** [review]

File: cmake/cuda.cmake (hunk 1 above)
Comment (yhmtsai) [reply cmake/cuda.cmake:25]: not sure about PGI.
For NVHPC, I will assume it is always shipped with the cuda toolkit

Comment (tcojean) [reply cmake/cuda.cmake:25]: the problem is the `nvc++` compiler is identified as `PGI` at least on Perlmutter...

File: omp/distributed/partition_kernels.cpp (hunk 3 above)
Comment (tcojean) [reply omp/distributed/partition_kernels.cpp:64]: fair enough I can remove this

**Review by pratikvn — APPROVED** [review]
LGTM!

**Pull request merged**